### PR TITLE
Refactor of PySRRegressor

### PR DESCRIPTION
**Pull request by tttc3**
Re Issue #143

Compatibility with scikit-learn should be improved. 

Noteable breaking changes for users: `PySRRegressor.equations` is now called `PySRRegressor.equations_`

Tests have been updated to allow compatibility with the refactored code but still assess the same functionality. All tests should pass.

Please let me know if there are any concerns or if you would like me to document/explain any of the changes in detail.

### CODE REVIEW

**Comment by MilesCranmer** [comment]
Awesome contribution, thank you so much @tttc3! This is great. I am reviewing now.

**Comment by MilesCranmer** [comment]
General comment: is there a way I can continue a search? In the current PySR, if you run `model.fit(X, y)`, and then `model.fit(X, y)`, a **second time**, it will continue where it left off. Is there a way to do that with this refactoring?

Basically, it stores the Julia output into `self.raw_julia_state_`. However, here it looks like it is set to `None` every call to `fit`. Maybe by default, it should leave it as `None`, or, alternatively, there could be a parameter passed to `fit` to "continue" the search?

**Comment by MilesCranmer** [comment]
I see this warning when I run `model.predict(X)` when `X` is a numpy array:

```
/Users/mcranmer/venvs/main/lib/python3.8/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but PySRRegressor was fitted with feature names
```

**Comment by MilesCranmer** [comment]
The following code produces an error with these changes when it doesn't on master. Something to do with feature selection?

```python
import numpy as np
import pandas as pd
from pysr import PySRRegressor

X = pd.DataFrame({f"k{i}": np.random.randn(1000) for i in range(30)})
y = X["k15"] ** 2 + np.cos(X["k20"])

model = PySRRegressor(unary_operators=["cos"], select_k_features=3)
model.fit(X, y)
ypred = model.predict(X)  # Errors
```

File: pysr/sr.py [hunk 1]
```diff
@@ -277,548 +187,610 @@ def __call__(self, X):
...
+         before passing to the symbolic regression code. None means no
+         feature selection; an int means select that many features.
+
+    kwargs : dict, default=None
+        Supports deprecated keyword arguments. Other arguments will
+        result in an error.
+
+    Attributes
+    ----------
+    equations_ : pandas.DataFrame
```
[the remark this answers is not in the document]
Comment (tttc3) [reply pysr/sr.py:500]: Yeah this is good with me!

File: pysr/sr.py [hunk 2]
```diff
@@ -277,548 +187,610 @@ def __call__(self, X):
...
-        weights=None,
-        binary_operators=None,
-        unary_operators=None,
-        procs=cpu_count(),
-        loss="L2DistLoss()",
-        complexity_of_operators=None,
-        complexity_of_constants=None,
-        complexity_of_variables=None,
-        populations=15,
+        binary_operators=[
```
[the remark this answers is not in the document]
Comment (tttc3) [reply pysr/sr.py:574]: I agree this is probably a good idea

File: test/test.py [hunk 3]
```diff
@@ -21,7 +22,7 @@ def setUp(self):
             inspect.signature(PySRRegressor.__init__).parameters["populations"].default
         )
         self.default_test_kwargs = dict(
-            model_selection="accuracy",
+            model_selection="best",
```
[the remark this answers is not in the document]
Comment (tttc3) [reply test/test.py:24]: I think I changed this while debugging something and forgot to revert it before commiting.

**Comment by tttc3** [comment]
The latest commit should resolve the following:
https://github.com/MilesCranmer/PySR/pull/146#discussion_r883979599 - Revert default values and perform validation
https://github.com/MilesCranmer/PySR/pull/146#issuecomment-1140150247 - Handle warning message properly
https://github.com/MilesCranmer/PySR/pull/146#issuecomment-1140158413 - Pass new unit-test

With respect to continuing a fit, the scikit-learn standard way is to inform the fit method via the `warm_start` parmater that is passed in init. If `warm_start` is `True` and `fit` is called a second time fitting continues from where it left off. I'll have a look at implmenting thing now.

**Comment by MilesCranmer** [comment]
Thanks!

I'm still seeing some issues with the selection for some reason. For example, if I take the above example, and pass a numpy array instead of a pandas dataframe:
```python
import numpy as np
import pandas as pd
from pysr import PySRRegressor

X = pd.DataFrame({f"k{i}": np.random.randn(1000) for i in range(30)})
y = X["k15"] ** 2 + np.cos(X["k20"])

model = PySRRegressor(unary_operators=["cos"], select_k_features=3)
model.fit(X.values, y.values)
ypred = model.predict(X.values)  # Errors
```

**Comment by MilesCranmer** [comment]
Hm, it's showing this:
```
AssertionError: Estimator PySRRegressor should not change or mutate  the parameter binary_operators from None to ['+', '*', '-', '/'] during fit.
```

Maybe this means that `None` needs to be converted to `['+', ...]` at `__init__`, rather than at `fit`?

**Comment by MilesCranmer** [comment]
Hmm. I think it should still be possible to get cross_val_score to work, but I realized that the **full** scikit-learn check_estimator test may be impossible to pass.

The reason is that some of the attributes of PySRRegressor are not pickleable. For example, `self.equations_` contains lambda functions, and `self.raw_julia_state_` is a PyCall object with some pointers which are impossible to pickle (since they require storing the entire runtime basically). 

I'm fine if the `check_estimator` test does not pass for now, as long as cross_val_score and other basic scikit-learn functionality seems to be working!

File: pysr/export_jax.py [hunk 4]
```diff
@@ -192,9 +192,6 @@ def sympy2jax(expression, symbols_in, selection=None, extra_jax_mappings=None):
     )
     hash_string = "A_" + str(abs(hash(str(expression) + str(symbols_in))))
     text = f"def {hash_string}(X, parameters):\n"
-    if selection is not None:
```
Comment (MilesCranmer) [inline pysr/export_jax.py:195]: Why was this part removed? It seems to make the JAX output no longer work when feature selection is turned on, e.g.,

```python
import numpy as np
import pandas as pd
from pysr import PySRRegressor
from functools import partial

X = pd.DataFrame({f"k{i}": np.random.randn(1000) for i in range(10, 21)})
y = X["k15"] ** 2 + np.cos(X["k20"])

model = PySRRegressor(unary_operators=["cos"], select_k_features=3, early_stop_condition=1e-5)
model.fit(X.values, y.values)
f, parameters = model.jax().values()

np_prediction = model.predict
jax_prediction = partial(f, parameters=parameters)

np_output = np_prediction(X.values)
jax_output = jax_prediction(X.values)

np_loss = np.average(np_output - y.values)
jax_loss = np.average(jax_output - y)
```

`np_loss` is zero but `jax_loss` seems to be much larger - indicating the input array is not being masked with the selection.

Comment (MilesCranmer) [reply pysr/export_jax.py]: Okay I reverted this and it seems to work!

**Comment by MilesCranmer** [comment]
I made it so that all the scikit-learn checks get printed individually. Just run:
```python
python -m unittest test.test.TestMiscellaneous.test_scikit_learn_compatibility
```
to see which ones pass/fail

**Comment by tttc3** [comment]
WIth respect to the pickling I was wondering if to drop raw_julia_state_ from the pickle.dumps() by using a custom `__getstate__` method and to print a warning stating that this was being done? I think the lambda issue can be avoided by defining _lambda as a property like this:
```Python
    @property
    def _lambda(self):
        return lambdify(self._sympy_symbols, self._sympy)
```

**Comment by tttc3** [comment]
For cleaning up the parameters that may be modified from init, perhaps they can still be validated in `_validate_params`, and then the modified values packed into a dictionary that is passed to `_run` and unpacked in there? Maybe something like this:

```Python
 def _validate_params(self):
    # These are the parameters which may be modified from the ones
    # specified in init, so we define them here locally:

    # Dict of parameter names and mutated defaults
    modifiable_params = {
        "binary_operators": "+ * - /".split(" "),
        "unary_operators": [],
        "constraints": {}
    }
    packed_modified_params = {}
    for parameter, default_value in modifiable_params.items():
        parameter_value = getattr(self, parameter)
        if parameter_value is None:
            parameter_value = default_value
        else:
            # Special cases such as when binary_operators is a string
            if parameter in ["binary_operators", "unary_operators"] and isinstance(parameter_value, str):
                parameter_value = [parameter_value]
        # Add to packed dict
        packed_modified_params[parameter] = parameter_value      
    return packed_modified_params
    
def _run(self, packed_modified_params):
    binary_operators=packed_modified_params["binary_operators"]
    unary_operators=packed_modified_params["unary_operators"]
    constraints=packed_modified_params["constraints"]
    
    # Run solver
```

These packed params could be stored as an attribute such as `mutated_input_params_`? I think this is also closer to original implementation.

File: pysr/sr.py [hunk 5]
```diff
@@ -1394,48 +1334,318 @@ def _run(self, X, y, weights, variable_names):
...
+        weights : {ndarray | pandas.DataFrame} of the same shape as y, default=None
+            Each element is how to weight the mean-square-error loss
+            for that particular element of y.
+
+        variable_names : list[str], default=None
+            A list of names for the variables, other than "x0", "x1", etc.
+            If :param`X` is a pandas dataframe, the column names will be used.
+            If variable_names are specified
+
+        from_equation_file : bool, default=False
```
[the remark this answers is not in the document]
Comment (tttc3) [reply pysr/sr.py:1376]: If checkpoints can only be loaded for fitted estimators, creating a new `load_from_checkpoint` or `load` function would be no issue.  Something like this would probably work nicely: 

However for unfitted estimators it is quite difficult as `get_hof` needs to know fit dependant information such as `selection_mask_`, `feature_names_in_`, and `nout_`. The current implmentation with `from_equation_file` was the simplest way around this I could think of. If feature selection didn't occur as part of the fit (feature selection was performed outside of the PySRRegressor), then `selection_mask_` is not needed and `feature_names_in_` could be overwritten by a parameter in a new `model.load_from_checkpoint` method.

**Comment by MilesCranmer** [comment]
The ideas you have for solving pickling sound like a great solution! Happy to have that. Also, for `__getstate__`, the columns `torch_format` and `jax_format` should be removed from `self.equations_`, if they exist. They can always be re-generated when the user calls `.pytorch()` or `.jax()`. (Admittedly, it might be worth it to refactor that in the future, but it's okay to leave it for now)

I'm also on board with the idea for `packed_modified_params`!

**Comment by MilesCranmer** [comment]
By the way - the various checks are here: https://github.com/scikit-learn/scikit-learn/blob/80598905e517759b4696c74ecc35c6e2eb508cff/sklearn/utils/estimator_checks.py#L3110 since the error messages from the unittest don't describe all the details of each test.

File: pysr/sr.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MilesCranmer) [reply pysr/sr.py:1376]: This sounds good to me!

**Comment by MilesCranmer** [comment]
Thanks for adding those changes!

Looks like most of the remaining issues are due to the fact that PySR is not deterministic. To enforce determinism I need to make some backend changes, and serial mode needs to be set in those tests, instead of the default multithreaded mode.

I think it is fine if those other tests fail for now–we can make the fails soft. The one remaining test which still looks to fail is `check_dict_unchanged`.

**Comment by MilesCranmer** [comment]
Thanks for finding this!! This is very helpful, and I actually had no idea that this was the cause for non-determinism. It looks like there is simply not enough resolution in the default `time()` call to use it as a birth order for equations:
```julia
import Base.Libc: TimeVal
println([TimeVal().usec for i in 1:10])
```
(several of these will be the same, indicating that there could be two equations with the same `.birth` - hence relying on the phase of the time code, which is unreliable...)

I will see if I can get this working in the backend.

**Comment by MilesCranmer** [comment]
Okay I added a `deterministic` parameter (added to new version of backend), which, when combined with `multithreading=False, procs=0, random_state=<int>`, gives deterministic results for PySR. Thanks for getting the other ones working. All the scikit-learn compatibility tests now pass!

**Comment by MilesCranmer** [comment]
It looks like it's mostly working, except for this weird `free(): invalid pointer` issue which appears on the linux test (and only sometimes)... Any idea where that came from?

**Comment by tttc3** [comment]
I believe it is an issue with pytorch/how it is compiled or imports shared libraries. There is no issue when using the pytorch version that is packaged with conda. Importing pytorch at the top of `export_pytorch.py`, rather than in the `_initialise_pytorch` function call also seemed to avoid the error. I think this works because pytorch loads the libraries before julia starts. Though I'm still not 100% exactly what is going on.

**Comment by MilesCranmer** [comment]
This is very strange. On macOS, importing torch *before* starting Julia causes the error:
```
Python(68571,0x1050ec580) malloc: Heap corruption detected, free list is damaged at 0x600002627f00
*** Incorrect guard value: 0
Python(68571,0x1050ec580) malloc: *** set a breakpoint in malloc_error_break to debug

signal (6): Abort trap: 6
in expression starting at none:0
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
Allocations: 10557118 (Pool: 10549797; Big: 7321); GC: 12
```
but on linux, it seems like the opposite order triggers the error...

I'm trying to make a MWE and will let you know

**Comment by MilesCranmer** [comment]
Okay here is a MWE:
```python
def do_something_julia():
    from julia import Pkg
    Pkg.activate("test", shared=True)
    Pkg.update()

def do_something_torch():
    import torch
```
If you run `do_something_torch` and then `do_something_julia`, I see a segfault. But the other order, it works.

Can you try on your system and see what order triggers an error?

**Comment by MilesCranmer** [comment]
Okay I got it working with the following "band-aid fix" to `test_torch.py`. This issues is deeper and not really part of this PR so we can fix it later.

```python
import platform

if platform.system() == "Darwin":  # (macOS)
    # Julia, then Torch
    from pysr.julia_helpers import init_julia

    Main = init_julia()
    import torch
else:
    # Torch, then Julia
    import torch
```

**Comment by MilesCranmer** [comment]
Okay looks like everything is working now. Let me know if you'd like to look over anything, otherwise I can merge.

Also added `early_stop_condition` to all the tests, which makes them finish earlier if an equation is found early in searches. This speeds up all the testing by quite a bit...

**Review by MilesCranmer — APPROVED** [review]
Looks fantastic! Thanks so much for this incredible contribution. I have reviewed all changes and it's looking great.

**Comment by tttc3** [comment]
Thanks, glad to help! Just addressed some DeepSource issues but other than that, hopefully everything is good for merge. I'm investigating the pytorch issue and will follow up anything useful in your issue.
